File: incubators/net.sf.j2s.java.org.eclipse.osgi/src/org/eclipse/core/runtime/internal/adaptor/ContextFinder.java
/*******************************************************************************
 * Copyright (c) 2005 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *******************************************************************************/
package org.eclipse.core.runtime.internal.adaptor;

import java.io.IOException;
import java.net.URL;
import java.security.AccessController;
import java.security.PrivilegedAction;
import java.util.Enumeration;

public class ContextFinder extends ClassLoader implements PrivilegedAction {

	private static final class Finder extends SecurityManager {
		public Class[] getClassContext() {
			return super.getClassContext();
		}
	}

	static final Finder contextFinder = (Finder) AccessController.doPrivileged(new PrivilegedAction() {
		public Object run() {
			return new Finder();
		}
	});

	public ContextFinder(ClassLoader contextClassLoader) {
		super(contextClassLoader);
	}

	// Return the first classloader of the stack that is neither the ContextFinder classloader nor the boot classloader.
	// We assume that the bootclassloader never uses the context classloader to find classes in itself.
	// If the classloader found is the parent then return null 
	ClassLoader basicFindClassLoader() {
		Class[] stack = contextFinder.getClassContext();
		ClassLoader result = null;
		for (int i = 1; i < stack.length; i++) {
			ClassLoader tmp = stack[i].getClassLoader();
			if (stack[i] != ContextFinder.class && tmp != null && tmp != this) {
				result = tmp;
				break;
			}
		}
		if (checkClassLoader(result))
			return result;
		return null;
	}

	private boolean checkClassLoader(ClassLoader classloader) {
		if (classloader == null || classloader == getParent())
			return false;
		for (ClassLoader parent = classloader.getParent(); parent != null; parent = parent.getParent())
			if (parent == this)
				return false;
		return true;
	}

	private ClassLoader findClassLoader() {
		if (System.getSecurityManager() == null)
			return basicFindClassLoader();
		return (ClassLoader) AccessController.doPrivileged(this);
	}

	public Object run() {
		return basicFindClassLoader();
	}

	protected synchronized Class loadClass(String arg0, boolean arg1) throws ClassNotFoundException {
		Class result = null;
		try {
			result = super.loadClass(arg0, arg1);
		} catch (ClassNotFoundException e) {
			//Ignore
		}
		if (result == null) {
			ClassLoader toConsult = findClassLoader();
			if (toConsult != null)
				result = toConsult.loadClass(arg0);
		}
		if (result == null)
			throw new ClassNotFoundException(arg0);
		return result;
	}

	protected URL findResource(String arg0) {
		URL result = super.findResource(arg0);
		if (result == null) {
			ClassLoader toConsult = findClassLoader();
			if (toConsult != null)
				result = toConsult.getResource(arg0);
		}
		return result;
	}

	protected Enumeration findResources(String arg0) {
		try {
			Enumeration result = super.findResources(arg0);
			if (result == null) {
				ClassLoader toConsult = findClassLoader();
				if (toConsult != null)
					result = toConsult.getResources(arg0);
			}
			return result;
		} catch (IOException e) {
			return null;
		}
	}
}

File: old/incubator/net.sf.j2s.test.mauve/j2slib/org/eclipse/core/runtime/internal/adaptor/ContextFinder.js
Clazz.declarePackage ("org.eclipse.core.runtime.internal.adaptor");
Clazz.load (["java.lang.ClassLoader", "$.SecurityManager", "java.security.PrivilegedAction", "$.AccessController"], "org.eclipse.core.runtime.internal.adaptor.ContextFinder", ["java.lang.ClassNotFoundException"], function () {
c$ = Clazz.declareType (org.eclipse.core.runtime.internal.adaptor, "ContextFinder", ClassLoader, java.security.PrivilegedAction);
Clazz.pu$h ();
c$ = Clazz.declareType (org.eclipse.core.runtime.internal.adaptor.ContextFinder, "Finder", SecurityManager);
c$ = Clazz.p0p ();
Clazz.defineMethod (c$, "basicFindClassLoader", 
function () {
var stack = org.eclipse.core.runtime.internal.adaptor.ContextFinder.contextFinder.getClassContext ();
var result = null;
for (var i = 1; i < stack.length; i++) {
var tmp = stack[i].getClassLoader ();
if (stack[i] !== org.eclipse.core.runtime.internal.adaptor.ContextFinder && tmp != null && tmp !== this) {
result = tmp;
break;
}}
if (this.checkClassLoader (result)) return result;
return null;
});
Clazz.defineMethod (c$, "checkClassLoader", 
($fz = function (classloader) {
if (classloader == null || classloader === this.getParent ()) return false;
for (var parent = classloader.getParent (); parent != null; parent = parent.getParent ()) if (parent === this) return false;

return true;
}, $fz.isPrivate = true, $fz), "ClassLoader");
Clazz.defineMethod (c$, "findClassLoader", 
($fz = function () {
if (System.getSecurityManager () == null) return this.basicFindClassLoader ();
return java.security.AccessController.doPrivileged (this);
}, $fz.isPrivate = true, $fz));
Clazz.overrideMethod (c$, "run", 
function () {
return this.basicFindClassLoader ();
});
Clazz.defineMethod (c$, "loadClass", 
function (arg0, arg1) {
var result = null;
try {
result = Clazz.superCall (this, org.eclipse.core.runtime.internal.adaptor.ContextFinder, "loadClass", [arg0, arg1]);
} catch (e) {
if (Clazz.instanceOf (e, ClassNotFoundException)) {
} else {
throw e;
}
}
if (result == null) {
var toConsult = this.findClassLoader ();
if (toConsult != null) result = toConsult.loadClass (arg0);
}if (result == null) throw  new ClassNotFoundException (arg0);
return result;
}, "~S,~B");
Clazz.defineMethod (c$, "findResource", 
function (arg0) {
var result = Clazz.superCall (this, org.eclipse.core.runtime.internal.adaptor.ContextFinder, "findResource", [arg0]);
if (result == null) {
var toConsult = this.findClassLoader ();
if (toConsult != null) result = toConsult.getResource (arg0);
}return result;
}, "~S");
Clazz.defineMethod (c$, "findResources", 
function (arg0) {
try {
var result = Clazz.superCall (this, org.eclipse.core.runtime.internal.adaptor.ContextFinder, "findResources", [arg0]);
if (result == null) {
var toConsult = this.findClassLoader ();
if (toConsult != null) result = toConsult.getResources (arg0);
}return result;
} catch (e) {
if (Clazz.instanceOf (e, java.io.IOException)) {
return null;
} else {
throw e;
}
}
}, "~S");
c$.contextFinder = c$.prototype.contextFinder = java.security.AccessController.doPrivileged ((function (i$, v$) {
if (!Clazz.isClassDefined ("org.eclipse.core.runtime.internal.adaptor.ContextFinder$1")) {
Clazz.pu$h ();
c$ = Clazz.declareAnonymous (org.eclipse.core.runtime.internal.adaptor, "ContextFinder$1", null, java.security.PrivilegedAction);
Clazz.overrideMethod (c$, "run", 
function () {
return  new org.eclipse.core.runtime.internal.adaptor.ContextFinder.Finder ();
});
c$ = Clazz.p0p ();
}
return Clazz.innerTypeInstance (org.eclipse.core.runtime.internal.adaptor.ContextFinder$1, i$, v$);
}) (this, null));
});
